lib/src/native/realm_core.dart

```diff
@@ -1509,80 +1495,64 @@ class _RealmCore {
+    final controller = userdata as SessionConnectionStateController;
 
     controller.onConnectionStateChange(ConnectionState.values[oldState], ConnectionState.values[newState]);
   }
 
   Future<void> sessionWaitForUpload(Session session) {
     final completer = Completer<void>();
-    _realmLib.realm_dart_sync_session_wait_for_upload_completion(
-      session.handle._pointer,
-      Pointer.fromFunction(_sessionWaitCompletionCallback),
-      completer.toPersistentHandle(),
-      _realmLib.addresses.realm_dart_delete_persistent_handle,
-      scheduler.handle._pointer,
-    );
+    final callback = Pointer.fromFunction<Void Function(Handle, Pointer<realm_sync_error_code_t>)>(_sessionWaitCompletionCallback);
+    final userdata = _realmLib.realm_dart_userdata_async_new(completer, callback.cast(), scheduler.handle._pointer);
+    _realmLib.realm_sync_session_wait_for_upload_completion(session.handle._pointer, _realmLib.addresses.realm_dart_sync_wait_for_completion_callback,
+        userdata.cast(), _realmLib.addresses.realm_dart_userdata_async_free);
     return completer.future;
   }
 
   Future<void> sessionWaitForDownload(Session session) {
     final completer = Completer<void>();
-    _realmLib.realm_dart_sync_session_wait_for_download_completion(
-      session.handle._pointer,
-      Pointer.fromFunction(_sessionWaitCompletionCallback),
-      completer.toPersistentHandle(),
-      _realmLib.addresses.realm_dart_delete_persistent_handle,
-      scheduler.handle._pointer,
-    );
+    final callback = Pointer.fromFunction<Void Function(Handle, Pointer<realm_sync_error_code_t>)>(_sessionWaitCompletionCallback);
+    final userdata = _realmLib.realm_dart_userdata_async_new(completer, callback.cast(), scheduler.handle._pointer);
+    _realmLib.realm_sync_session_wait_for_download_completion(session.handle._pointer, _realmLib.addresses.realm_dart_sync_wait_for_completion_callback,
+        userdata.cast(), _realmLib.addresses.realm_dart_userdata_async_free);
     return completer.future;
   }
 
-  static void _sessionWaitCompletionCallback(Pointer<Void> userdata, Pointer<realm_sync_error_code_t> errorCode) {
-    try {
-      final completer = userdata.toObject<Completer<void>>(isPersistent: true);
-      if (completer == null) {
-        return;
-      }
+  static void _sessionWaitCompletionCallback(Object userdata, Pointer<realm_sync_error_code_t> errorCode) {
+    final completer = userdata as Completer<void>;
 
-      if (errorCode != nullptr) {
-        // Throw RealmException instead of RealmError to be recoverable by the user.
-        completer.completeError(RealmException(errorCode.toSyncError().toString()));
-      } else {
-        completer.complete();
-      }
-    } finally {
-      _realmLib.realm_free(errorCode.cast());
+    if (errorCode != nullptr) {
+      completer.completeError(errorCode.toSyncError());
+    } else {
+      completer.complete();
```

Comment:
Are we sure `realm_free` is not needed here

Reply:
The string memory is owned by the `realm_dart_sync_wait_for_completion_callback` trampoline, we don't need to free it in Dart.

Reply:
leaving open just to confirm it is owned in the native struct copy